correct, cleanup & simplify dwim stack

bcoca opened the pull request:
##### SUMMARY
- minor improvements to 'abs path detection'
- use unfrackpath instead of recreating it X times
- try to improve role detection

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Bugfix Pull Request

##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->
dwim

##### ANSIBLE VERSION
<!--- Paste verbatim output from “ansible --version” between quotes below -->
```
2.4
```

bcoca committed:
correct, cleanup & simplify dwim stack

latlh chIS logh HeS qar wej chel laD
better errors
update find_file to new exception

mattclay commented on lib/ansible/errors/__init__.py:
```diff
@@ -220,7 +220,24 @@ class AnsibleUndefinedVariable(AnsibleRuntimeError):
+            message += "Could not find or access '%s'" % to_text(self.file_name)
+
+        if self.paths and isinstance(self.paths, list):
+            searched = to_text('\n\t'.join(self.file_paths)
```
CI failure due to syntax error here.

mattclay commented:
I've added the missing `__init__.py` file. The `validate-modules` test missed it on the PR that created the directory. I'll have a fix for that soon.

mattclay commented:
CI failure in integration tests. Here's one of the errors:

```
Traceback (most recent call last):
  File "/root/ansible/lib/ansible/executor/task_executor.py", line 88, in run
    items = self._get_loop_items()
  File "/root/ansible/lib/ansible/executor/task_executor.py", line 216, in _get_loop_items
    items = mylookup.run(terms=loop_terms, variables=self._job_vars, wantlist=True)
  File "/root/ansible/lib/ansible/plugins/lookup/first_found.py", line 185, in run
    path = self.find_file_in_search_path(variables, subdir, fn, ignore_missing=True)
  File "/root/ansible/lib/ansible/plugins/lookup/__init__.py", line 118, in find_file_in_search_path
    result = self._loader.path_dwim_relative_stack(paths, subdir, needle)
  File "/root/ansible/lib/ansible/parsing/dataloader.py", line 378, in path_dwim_relative_stack
    raise AnsibleFileNotFound(file_name=source, paths=search)
  File "/root/ansible/lib/ansible/errors/__init__.py", line 235, in __init__
    searched = to_text('\n\t'.join(self.file_paths))
AttributeError: 'AnsibleFileNotFound' object has no attribute 'file_paths'
```

mattclay commented:
CI failure in windows integration test `test/integration/targets/win_copy/tasks/main.yml:327`:

```
{
"changed": false, 
"failed": true, 
"failed_when_result": true, 
"msg": "Could not find or access 'false-file'\nSearched in:\n\t/root/src/github.com/ansible/ansible/test/integration/targets/win_copy/files/false-file\n\t/root/src/github.com/ansible/ansible/test/integration/targets/win_copy/false-file\n\t/root/src/github.com/ansible/ansible/test/integration/targets/win_copy/tasks/files/false-file\n\t/root/src/github.com/ansible/ansible/test/integration/targets/win_copy/tasks/false-file\n\t/root/src/github.com/ansible/ansible/test/integration/files/false-file\n\t/root/src/github.com/ansible/ansible/test/integration/false-file"
}
```

bcoca commented:
@mattclay error message changed and the test looks for exact match ...

alikins commented on lib/ansible/errors/__init__.py:
```diff
@@ -220,7 +220,24 @@ class AnsibleUndefinedVariable(AnsibleRuntimeError):
+                message += "\n"
+            message += "Searched in:\n\t%s" % searched
+
+        super(AnsibleFileNotFound, self).__init__(message=message, obj=obj, show_content=show_content, orig_exc=orig_exc)
```
You are suppressing the suppress_extended_error arg if used when creating a AnsibleFileNotFound()

alikins commented on lib/ansible/errors/__init__.py:
```diff
@@ -220,7 +220,24 @@ class AnsibleUndefinedVariable(AnsibleRuntimeError):
+            searched = to_text('\n\t'.join(self.paths))
+            if message:
+                message += "\n"
+            message += "Searched in:\n\t%s" % searched
```
If I were writing the exceptions from scratch, I think I would tend to defer most of the message formatting to the __repr__/__str__ if possible. But I'm not sure how intertwined the the message creation in the base AnsibleError __init__ is required, so there may be reasons we can't defer to __repr__...   ie, if there is code explicitly using AnsibleError.message instead of the str/repr results 

It looks like there are some references to some_exception.message, though at first inspection I dont see any that would do that for an AnsibleError, so moving that to the repr/str might work.

bcoca replied:
i didn't want to redo everything that is already there, but does sound like a more pythonic approach

alikins commented on lib/ansible/parsing/dataloader.py:
```diff
@@ -184,7 +187,7 @@ def _get_file_contents(self, file_name):
         b_file_name = to_bytes(file_name)
         if not self.path_exists(b_file_name) or not self.is_file(b_file_name):
-            raise AnsibleFileNotFound("the file named '%s' does not exist, or is not readable" % file_name)
+            raise AnsibleFileNotFound("Unable to retrieve file contents", file_name=file_name)
```
With the AnsibleFileNotFound changes, not sure we need to specify a message here.

bcoca replied:
I left it so we could if we want to, since this one is about file contents i think it is more pertinent than the default

alikins commented on lib/ansible/errors/__init__.py:
```diff
@@ -220,7 +220,24 @@ class AnsibleUndefinedVariable(AnsibleRuntimeError):
     ''' a file missing failure '''
-    pass
+
+    def __init__(self, message="", obj=None, show_content=True, suppress_extended_error=False, orig_exc=None, paths=None, file_name=None):
```
maybe s/paths/searched_paths to distinguish from 'file_name' a little more? Though the message formatting can tweak that however for using facing messages so doesnt matter a ton.

alikins commented on lib/ansible/parsing/dataloader.py:
```diff
@@ -273,9 +274,10 @@ def path_dwim_relative(self, path, dirname, source, is_role=False):
 
         # I have full path, nothing else needs to be looked at
-        if source.startswith('~') or source.startswith(os.path.sep):
+        if RE_ABS_PATH.search(source):
```
RE_ABS_PATH is maybe a little too concrete. Maybe wrap in a is_abs_path() method? ditto for RE_TASKS

alikins commented on lib/ansible/utils/path.py:
```diff
@@ -18,13 +18,16 @@
-__all__ = ['unfrackpath', 'makedirs_safe']
+__all__ = ['unfrackpath', 'makedirs_safe', 'RE_ABS_PATH']
+
+RE_ABS_PATH = re.compile('^(?:~|%s)' % (os.path.sep))
```
I recommend adding some unit tests to directly test any defined regexes. Just makes it easier to test if you have to change it (and also provides example hits/misses). Granted, this is not a complex regex...

alikins reviewed (commented):
Looks like a good idea to me. A few exception tweaks here and there...

alikins commented on lib/ansible/plugins/lookup/__init__.py:
```diff
@@ -112,8 +113,11 @@ def find_file_in_search_path(self, myvars, subdir, needle, ignore_missing=False)
-        result = self._loader.path_dwim_relative_stack(paths, subdir, needle)
-        if result is None and not ignore_missing:
-            self._display.warning("Unable to find '%s' in expected paths." % needle)
+        result = None
```
[just a note inspired by branch, not something that needs to be addressed by this pr]

Looking though code that uses find_file_in_search_path(), there are some lookup plugins that dont really
check the return value (ini looks like it can throw exceptions if result is None, csvfile doesnt check for !None, but does at least try/except around open() it)

bcoca replied:
I was planning on fixing that in subsequent Prs, need to go over many of em

alikins commented:
[More notes that dont need to be addressed by this pr but noticed while reviewing]

There are some places that currently raise AnsibleError that look like a AnsibleFileNotFound might be more approriate:

```
lib/ansible/utils/hashing.py: raise AnsibleError("error while accessing the file %s, error was: %s" % (filename, e)
lib/ansible/parsing/vault/__init__.py: raise AnsibleError("The vault password file %s was not found" % this_path)
lib/ansible/parsing/vault/__init__.py: raise AnsibleError("Could not read vault password file %s: %s" % (this_path, e))
lib/ansible/plugins/lookup/template.py: raise AnsibleError("the template file %s could not be found for the lookup" % term
lib/ansible/plugins/connection/zone.py: raise AnsibleError("file or module does not exist at: %s" % in_path)
lib/ansible/plugins/connection/chroot.py: raise AnsibleError("file or module does not exist at: %s" % 
lib/ansible/plugins/connection/jail.py: raise AnsibleError("file or module does not exist at: %s" % in_path)
lib/ansible/plugins/lookup/shelvefile.py: raise AnsibleError("Could not locate shelve file in lookup: %s" % file)
lib/ansible/cli/config.py: raise AnsibleError("Failed to open config file: %s" % to_native(e))
lib/ansible/cli/galaxy.py: raise AnsibleError('Unable to open %s: %s' % (role_file, str(e)))
```

abadger commented on lib/ansible/errors/__init__.py:
```diff
@@ -220,7 +220,24 @@ class AnsibleUndefinedVariable(AnsibleRuntimeError):
+                message += "\n"
+            message += "Could not find or access '%s'" % to_text(self.file_name)
+
+        if self.paths and isinstance(self.paths, list):
```
* Need to also handle self.paths and not isinstance(self.paths, list)
* Use collections.Sequence in isinstance, that way both tuples and lists will work.

bcoca replied:
i was handling it by 'ignoring it as bad data'

abadger commented on lib/ansible/parsing/dataloader.py:
```diff
@@ -233,35 +236,33 @@ def path_dwim(self, given):
-            return os.path.abspath(given)
-        elif given.startswith(u"~"):
-            return os.path.abspath(os.path.expanduser(given))
+        if RE_ABS_PATH.search(given):
```
Don't use the regex here.  Several problems with it:
* regexes are slow.  Anytime you can replace a regex with a simple substring match use the substring functions.
* the functions above take care to use text strings.  The regex uses native strings.  That means on python2 you could have issues if nonascii characters creep in.

bcoca replied:
at which threshold is the regex better? these are 2 substring matches for each

abadger replied:
it depends on the data.  For regex, the longer the string the quicker performance drops off.  here's some testing:
``` python
>>> RE = re.compile('(?:a|b)')
>>> def match(test):
...  if test.endswith('a') or test.endswith('b'): return test
... 
>>> def match2(test):
...  if RE.search(test): return test
>>> timeit.repeat('match("1" * 1)', 'from __main__ import match')
[0.321303129196167, 0.3136630058288574, 0.31526708602905273]
>>> timeit.repeat('match2("1" * 1)', 'from __main__ import match2')
[0.32927894592285156, 0.3168821334838867, 0.31973695755004883]
>>> timeit.repeat('match("1" * 10)', 'from __main__ import match')
[0.31588292121887207, 0.3094758987426758, 0.31125402450561523]
>>> timeit.repeat('match2("1" * 10)', 'from __main__ import match2')
[0.40798497200012207, 0.4003629684448242, 0.3933870792388916]
>>> timeit.repeat('match("1" * 100)', 'from __main__ import match')
[0.37717509269714355, 0.3657679557800293, 0.3643019199371338]
>>> timeit.repeat('match2("1" * 100)', 'from __main__ import match2')
[1.2596900463104248, 1.2499568462371826, 1.239171028137207]
```

Unless your string sizes are very small, the rule of thumb should probably be that regex is never faster than any number of substring matches.

bcoca replied:
restored for the abs path, but kept tasks, that is waaay too many ifs to get same filter

abadger commented on lib/ansible/parsing/dataloader.py:
```diff
@@ -233,35 +236,33 @@ def path_dwim(self, given):
-            return os.path.abspath(os.path.join(basedir, given))
+            path = unfrackpath(os.path.join(basedir, given))
+
+        return path
```
Are we sure we want path_dwim to call unfrackpath?  That changes the semantics of path_dwim and could break code if it relied on symlinks not being derefernced (for instance).  Perhaps path_dwim() should call unfrackpath with follow=False?

bcoca replied:
it was already doing essentially the same in path_dwim, the only difference is that $VARS are now also allowed in relative paths

abadger replied:
This code is also more clearly written as:
``` python
if not (given.startswith(u"/") or given.startswith(u"~")):
    basedir = to_text(self._basedir, errors='surrogate_or_strict')
    given = os.path.join(basedir, given)

path = unfrackpath(given, follow=False)

return path
```

bcoca replied:
doesn't this have the same 'type problem' as the regex?

abadger replied:
* The path_dwim() code was not doing the same thing as ```unfrackpath([...], follow=True)```.  The path_dwim() code was doing approximately the same things as ```unfrackpath([...],  follow=False)```.
* The type problem with the regex is that, on python2 it mixes a bytes regex with text data.  path_dwim() currently expects text strings so we need to use text everywhere to avoid mixing types.

abadger commented on lib/ansible/parsing/dataloader.py:
```diff
@@ -233,35 +236,33 @@ def path_dwim(self, given):
 
-        return isit
+                if (
+                    RE_TASKS.search(path) and
```
I think the same comment about using substring matching instead of a regex applies here as it did earlier.  however, the regex looks different from the substring match it's replacing.  Can you tell me why that is and what it's supposed to be matching now?

bcoca replied:
the regex is more complete,  i.e it will match `tasks` and `tasks/` and `/tasks` and `afd/tasks/` but not ` sdlkftasks`

abadger commented on lib/ansible/parsing/dataloader.py:
```diff
@@ -233,35 +236,33 @@ def path_dwim(self, given):
-            return os.path.abspath(given)
-        elif given.startswith(u"~"):
-            return os.path.abspath(os.path.expanduser(given))
+        if given.matches(os.path.sep) or given.matches(u'~'):
```
You want ```.startswith()``` and instead of os.path.sep, ```to_text(os.path.sep)```

bcoca replied:
erght ... fixed ...

abadger commented on lib/ansible/parsing/dataloader.py:
```diff
@@ -233,35 +236,33 @@ def path_dwim(self, given):
             basedir = to_text(self._basedir, errors='surrogate_or_strict')
-            return os.path.abspath(os.path.join(basedir, given))
+            path = unfrackpath(os.path.join(basedir, given), follow=False)
+
```
The clearer that I was talking about earlier was that the unfrackpath is called in both branches of the conditional.  So it's clearer if that happens after exiting the conditional.

bcoca replied:
dones

abadger commented on lib/ansible/parsing/dataloader.py:
```diff
@@ -233,35 +236,33 @@ def path_dwim(self, given):
+                b_tasked = b'%s/%s' % (finddir, b_main)
+
+                if (
+                    RE_TASKS.search(path) and
```
* Should add a comment here saying what strings you intend to match.
* Could be done with a substring which would be quicker but it's less clear to read so I'm not considering a substring a blocker here. substring would look like this:
``` python
idx = path.rfind(u'tasks')
if (
    (idx == 0 or path[idx-1] == u'/') and idx + 5 == len(path) or (idx + 6 == len(path) and path[idx+5] == u'/') and
    [...]
```

abadger replied:
Since the new code uses RE_TASKS in other places, probably put the comment about what the intended matches should be at the top where RE_TASKS is defined.

abadger commented on lib/ansible/parsing/dataloader.py:
```diff
@@ -321,24 +322,24 @@ def path_dwim_relative_stack(self, paths, dirname, source, is_role=False):
         :arg source: A text string which is the filename to search for
         :rtype: A text string
-        :returns: An absolute path to the filename ``source``
+        :returns: An absolute path to the filename ``source`` or raises an AnsibleFileNotFound Exception
```
sphinx docstrings can have (multiple) raises fields:
```
:returns: An absolute path to the filename ``source``
:raises AnsibleFileNotFound: when the ``source`` was not found in any of the search paths.
```

abadger commented on lib/ansible/parsing/dataloader.py:
```diff
@@ -420,7 +424,7 @@ def get_real_file(self, file_path, decrypt=True):
                         data = f.read()
                         if not self._b_vault_password:
-                            raise AnsibleParserError("A vault password must be specified to decrypt %s" % file_path)
+                            raise AnsibleParserError("A vault password must be specified to decrypt %s" % to_text(file_path))
```
This is where the variable is being handed to the message of an exception constructor.  So it's where to_native() is appropriate.

abadger commented on lib/ansible/utils/path.py:
```diff
@@ -18,6 +18,7 @@
 __metaclass__ = type
 
 import os
+import re
```
I think we don't need the re import anymore as we're no longer defining a regex here.

abadger commented:
Okay, just a few more comments.  once those are addressed, +1 to merge from me.

bcoca committed:
addressed latest comments

bcoca committed:
test should not use realpath as it follows symlink

this fails when on OS X as /var is now a symlink to /private/var
but first_found was not supposed to follow symlinks
